### PR TITLE
feat: add tx cli to build evm transaction

### LINKED ISSUE #709: No tx cli to build evm transaction
__Proposal:__ No tx command to build evm transaction

__Current behavior:__ only web3 API to build and broadcast transaction

__Desired behavior:__ Implement tx command

__Use case:__

- generate_only
- broadcast with different mode

### PR DESCRIPTION
**Pull request by yihuang**
Closes #709

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/client/cli/tx.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	cmd := &cobra.Command{
+		Use:                        types.ModuleName,
+		Short:                      fmt.Sprintf("%s transactions subcommands", types.ModuleName),
+		DisableFlagParsing:         true,
+		SuggestionsMinimumDistance: 2,
+		RunE:                       client.ValidateCmd,
+	}
+
+	// this line is used by starport scaffolding # 1
+
```
Comment (fedekunze) [inline x/evm/client/cli/tx.go:31]:
```suggestion
```

File: x/evm/client/cli/tx.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	cmd.AddCommand(CmdRaw())
+	return cmd
+}
+
+// CmdRaw command build cosmos transaction from raw ethereum transaction
+func CmdRaw() *cobra.Command {
```
Comment (fedekunze) [inline x/evm/client/cli/tx.go:37]:
```suggestion
// NewRawTxCmd command build cosmos transaction from raw ethereum transaction
func NewRawTxCmd() *cobra.Command {
```

File: x/evm/client/cli/tx.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	cmd := &cobra.Command{
+		Use:   "raw [tx-hex]",
+		Short: "Build cosmos transaction from raw ethereum transaction",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			data, err := hexutil.Decode(args[1])
+			if err != nil {
+				return err
```
Comment (fedekunze) [inline x/evm/client/cli/tx.go:45]:
```suggestion
				return errors.Wrap(err, "failed to decode ethereum tx hex bytes")
```

File: x/evm/client/cli/tx.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+					return err
+				}
+
+				_, _ = fmt.Fprintf(os.Stderr, "%s\n\n", out)
+
+				buf := bufio.NewReader(os.Stdin)
+				ok, err := input.GetConfirmation("confirm transaction before signing and broadcasting", buf, os.Stderr)
+
+				if err != nil || !ok {
+					_, _ = fmt.Fprintf(os.Stderr, "%s\n", "cancelled transaction")
```
Comment (fedekunze) [inline x/evm/client/cli/tx.go:94]:
```suggestion
					_, _ = fmt.Fprintf(os.Stderr, "%s\n", "canceled transaction")
```

File: x/evm/types/msg.go [hunk 5]
```diff
@@ -282,3 +282,15 @@ func (msg *MsgEthereumTx) GetSender(chainID *big.Int) (common.Address, error) {
 func (msg MsgEthereumTx) UnpackInterfaces(unpacker codectypes.AnyUnpacker) error {
 	return unpacker.UnpackAny(msg.Data, new(TxData))
 }
+
+// UnmarshalBinary decodes the canonical encoding of transactions.
+func (msg *MsgEthereumTx) UnmarshalBinary(b []byte) error {
```
Comment (fedekunze) [inline x/evm/types/msg.go:287]: we should add a unit test for this

File: x/evm/types/msg.go [hunk 6]
```diff
@@ -282,3 +282,15 @@ func (msg *MsgEthereumTx) GetSender(chainID *big.Int) (common.Address, error) {
...
+	tx := &ethtypes.Transaction{}
+	if err := tx.UnmarshalBinary(b); err != nil {
+		return err
+	}
+	if err := msg.FromEthereumTx(tx); err != nil {
+		return err
+	}
+	return nil
```
Comment (fedekunze) [inline x/evm/types/msg.go:295]:
```suggestion
	return msg.FromEthereumTx(tx)
```

File: x/evm/client/cli/tx.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			data, err := hexutil.Decode(args[1])
+			if err != nil {
+				return err
+			}
+
+			msg := &types.MsgEthereumTx{}
+			if err := msg.UnmarshalBinary(data); err != nil {
+				return err
+			}
```
Comment (fedekunze) [inline x/evm/client/cli/tx.go:51]:
```suggestion
			if err := msg.UnmarshalBinary(data); err != nil {
				return err
			}
			
			if err := msg.ValidateBasic(); err != nil {
				return err
			}
```

File: x/evm/types/utils.go [hunk 8]
```diff
@@ -103,3 +106,32 @@ func SafeNewIntFromBigInt(i *big.Int) (sdk.Int, error) {
 func IsValidInt256(i *big.Int) bool {
 	return i == nil || i.BitLen() <= maxBitLen
 }
+
+// BuildRawTransaction build cosmos transaction from raw ethereum transaction
+func BuildRawTransaction(builder authtx.ExtensionOptionsTxBuilder, msg *MsgEthereumTx, evmDenom string) (signing.Tx, error) {
```
Comment (fedekunze) [inline x/evm/types/utils.go:111]: ditto unit test

File: x/evm/client/cli/tx.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// CmdRaw command build cosmos transaction from raw ethereum transaction
+func CmdRaw() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "raw [tx-hex]",
+		Short: "Build cosmos transaction from raw ethereum transaction",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			data, err := hexutil.Decode(args[1])
```
Comment (fedekunze) [inline x/evm/client/cli/tx.go:43]:
```suggestion
			data, err := hexutil.Decode(args[0])
```

**Review by fedekunze — CHANGES_REQUESTED** [review]
Looks good. Thanks, @yihuang. See comments below

File: x/evm/types/utils.go (hunk 8 above)
Comment (yihuang) [reply x/evm/types/utils.go]: done, also renamed to `BuildTx` method.

File: x/evm/types/msg.go (hunk 5 above)
Comment (yihuang) [reply x/evm/types/msg.go:291]: done

**Review by fedekunze — APPROVED** [review]
ACK. Pending Changelog entry and lint fixes

**Comment by yihuang** [comment]
> ACK. Pending Changelog entry and lint fixes

done

**Pull request merged**